Modified pull request removing large number of example config files in Lambda Support.

JPHammonds opened the pull request:
I have pulled the entire config directory from LambdaSuppport and left a text file pointing to a file on our confluence page which contains these files.  I have left a copy in the ioc directory, removing two copies that are not used.  I have also left one copy in the test directory.

MarkRivers commented:
The LambdaSupport directory cannot be built with the EPICS build system, it can only be built with cmake.  This is not the way we have done other detectors.  
It looks to me like the only things that really need to be in the LamdaSupport directory are fewer than 10 files each in libcommon/include, libcommon/src, liblambda/include, and liblamda/src.  It should be easy to add Makefiles in the LambdaSupport tree that will allow the required libraries to be built and include files installed with the EPICS build system.   
Do we need the libcommon/test and liblambda/test directories?  Without them the LamdaSupport directory shrinks from 19MB to less than 1 MB.
In this directory are those large configuration files generic, or are they specific to a particular serial number detector?
    ADLambda/iocs/LambdaIOC/iocBoot/iocLambda/config/FullRXModule14/lookups
    corvette:config/FullRXModule14/lookups>du -sh *
    3.1M    correction_si_1556_516.int32
    3.1M    nominator_si_1556_516.int32
    3.1M    pixelmask_si_1556_516.uint32

JPHammonds commented:
Mark,

Although we could recreate the build in Make, this would require 
tracking changes made by X-spectrum as the library changes.  I would 
rather use the build system created by the vendor.  I have included the 
library files here for convenience.  The README files at the top level 
point to instructions for building the library in a way that things will 
be found when building the ADLambda driver.  It is not exactly like 
other support directories, but those commonly use prebuilt libraries.

I have left the test directories since these allow distribution of the 
full source from the vendor.  I see how pulling the main config files is 
an issue. As far as I know the configuration files are fairly generic.  
They would at least provide a start. Users will need to modify the main 
file SystemConfig.txt since the MAC address defined for the host machine 
must match the actual machine.  I also imagine users will need config 
files tweaked for use with a particular detector.

John

On 4/13/2016 10:13 AM, Mark Rivers wrote:
>
> The LambdaSupport directory cannot be built with the EPICS build 
> system, it can only be built with cmake. This is not the way we have 
> done other detectors.
>
> It looks to me like the only things that really need to be in the 
> LamdaSupport directory are fewer than 10 files each in 
> libcommon/include, libcommon/src, liblambda/include, and liblamda/src. 
> It should be easy to add Makefiles in the LambdaSupport tree that will 
> allow the required libraries to be built and include files installed 
> with the EPICS build system.
>
> Do we need the libcommon/test and liblambda/test directories? Without 
> them the LamdaSupport directory shrinks from 19MB to less than 1 MB.
> In this directory are those large configuration files generic, or are 
> they specific to a particular serial number detector?
> ADLambda/iocs/LambdaIOC/iocBoot/iocLambda/config/FullRXModule14/lookups
> corvette:config/FullRXModule14/lookups>du -sh *
> 3.1M correction_si_1556_516.int32
> 3.1M nominator_si_1556_516.int32
> 3.1M pixelmask_si_1556_516.uint32
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-209500701>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

JPHammonds commented:
Mark,

Note that we also had to fight enough differences in machines used at 
DESY and here at Argonne, I would not want to have a change in how the 
libraries are built become an issue when discussing problems with the 
vendor.

John

On 4/13/2016 10:13 AM, Mark Rivers wrote:
>
> The LambdaSupport directory cannot be built with the EPICS build 
> system, it can only be built with cmake. This is not the way we have 
> done other detectors.
>
> It looks to me like the only things that really need to be in the 
> LamdaSupport directory are fewer than 10 files each in 
> libcommon/include, libcommon/src, liblambda/include, and liblamda/src. 
> It should be easy to add Makefiles in the LambdaSupport tree that will 
> allow the required libraries to be built and include files installed 
> with the EPICS build system.
>
> Do we need the libcommon/test and liblambda/test directories? Without 
> them the LamdaSupport directory shrinks from 19MB to less than 1 MB.
> In this directory are those large configuration files generic, or are 
> they specific to a particular serial number detector?
> ADLambda/iocs/LambdaIOC/iocBoot/iocLambda/config/FullRXModule14/lookups
> corvette:config/FullRXModule14/lookups>du -sh *
> 3.1M correction_si_1556_516.int32
> 3.1M nominator_si_1556_516.int32
> 3.1M pixelmask_si_1556_516.uint32
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-209500701>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

MarkRivers commented:
I just tried to run cmake following the instructions in INSTALL and it failed.

I ran these commands
>
/home/epics/devel/areaDetector/ADLambda/LambdaSupport
corvette:areaDetector/ADLambda/LambdaSupport>mkdir liblambda-build 
corvette:areaDetector/ADLambda/LambdaSupport>cd liblambda-build/
corvette:ADLambda/LambdaSupport/liblambda-build>cmake ../liblambda/src/
-- The C compiler identification is GNU
-- The CXX compiler identification is GNU
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
CMake Error at CMakeLists.txt:14 (install):
  install TARGETS given no LIBRARY DESTINATION for shared library target
  "Project".
CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as
    cmake_minimum_required(VERSION 2.8)
  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.
-- Configuring incomplete, errors occurred!

Can you add an INSTALL-EPICS file that gives an example of running cmake that works?  Where do you expect that cmake puts its output files?  The default according to INSTALL is /usr/local/.  Will your driver find the include and library files there?  It does not look like it, since LambdaApp/src/Makefile uses LIB_LIBS=lambda, so it is not looking for lamda in the system libraries, but rather in the EPICS application directories.

JPHammonds commented:
I have added a file in the LambdaSupport directory.  I thought I had 
enough information combining the README in LambdaSupport and INSTALL in 
liblambda.  I can see the confusion.  Hopefully this will help.

The cmake, make install inside the liblambda-build installs everything 
within this directory.  The make in LambdaSupport should install the 
headers and libraries in the correct place.  This of course will fail if 
the cmake stuff is not done.  It will also not pick up any changes to 
the driver without re-running cmake, etc.

John

On 4/13/2016 12:04 PM, Mark Rivers wrote:
>
> I just tried to run cmake following the instructions in INSTALL and it 
> failed.
>
> I ran these commands
>
>     /home/epics/devel/areaDetector/ADLambda/LambdaSupport
>     corvette:areaDetector/ADLambda/LambdaSupport>mkdir liblambda-build
>     corvette:areaDetector/ADLambda/LambdaSupport>cd liblambda-build/
>     corvette:ADLambda/LambdaSupport/liblambda-build>cmake
>     ../liblambda/src/
>     -- The C compiler identification is GNU
>     -- The CXX compiler identification is GNU
>     -- Check for working C compiler: /usr/bin/gcc
>     -- Check for working C compiler: /usr/bin/gcc -- works
>     -- Detecting C compiler ABI info
>     -- Detecting C compiler ABI info - done
>     -- Check for working CXX compiler: /usr/bin/c++
>     -- Check for working CXX compiler: /usr/bin/c++ -- works
>     -- Detecting CXX compiler ABI info
>     -- Detecting CXX compiler ABI info - done
>     CMake Error at CMakeLists.txt:14 (install):
>     install TARGETS given no LIBRARY DESTINATION for shared library target
>     "Project".
>     CMake Warning (dev) in CMakeLists.txt:
>     No cmake_minimum_required command is present. A line of code such as
>     cmake_minimum_required(VERSION 2.8)
>     should be added at the top of the file. The version specified may
>     be lower
>     if you wish to support older CMake versions for this project. For more
>     information run "cmake --help-policy CMP0000".
>     This warning is for project developers. Use -Wno-dev to suppress it.
>     -- Configuring incomplete, errors occurred!
>
> Can you add an INSTALL-EPICS file that gives an example of running 
> cmake that works? Where do you expect that cmake puts its output 
> files? The default according to INSTALL is /usr/local/. Will your 
> driver find the include and library files there? It does not look like 
> it, since LambdaApp/src/Makefile uses LIB_LIBS=lambda, so it is not 
> looking for lamda in the system libraries, but rather in the EPICS 
> application directories.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-209547326>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

MarkRivers commented:
I ran to a problem with my version of boost.  It was fixed by the tip at the end of this post

[https://bbs.archlinux.org/viewtopic.php?id=144593](https://bbs.archlinux.org/viewtopic.php?id=144593)

I then ran into this error:
corvette:ADLambda/LambdaSupport/liblambda-build>make
[ 10%] Building CXX object src/CMakeFiles/lambda.dir/ImageDecoder.cpp.o
In file included from /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda/src/ImageDecoder.cpp:22:0:
/home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda/include/ImageDecoder.h:26:23: fatal error: stdalign.h: No such file or directory
compilation terminated.

I am running gcc 4.6.3, while 4.4.7 listed in  liblambda/INSTALL.  stdalign.h apparently does not exist in 4.6.3.

MarkRivers commented:
I worked around the missing stdalign.h by just commenting out the line that includes it in ImageDecoder.h.  It compiled fine with no warnings.  I then rebuild everything, but I get an error when I do "make install"
corvette:ADLambda/LambdaSupport/liblambda-build>cmake --DMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
-- The C compiler identification is GNU
-- The CXX compiler identification is GNU
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Install directory is: /usr/local
-- Boost version: 1.46.0
-- No build type selected, default to RELEASE
-- Build type: RELEASE
-- Configuring done
-- Generating done
-- Build files have been written to: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-build
corvette:ADLambda/LambdaSupport/liblambda-build>make
Scanning dependencies of target lambda
[ 10%] Building CXX object src/CMakeFiles/lambda.dir/home/epics/devel/areaDetector/ADLambda/LambdaSupport/libcommon/src/Utils.cpp.o
[ 20%] Building CXX object src/CMakeFiles/lambda.dir/home/epics/devel/areaDetector/ADLambda/LambdaSupport/libcommon/src/NetworkInterface.cpp.o
[ 30%] Building CXX object src/CMakeFiles/lambda.dir/home/epics/devel/areaDetector/ADLambda/LambdaSupport/libcommon/src/NetworkImplementation.cpp.o
[ 40%] Building CXX object src/CMakeFiles/lambda.dir/home/epics/devel/areaDetector/ADLambda/LambdaSupport/libcommon/src/FilesOperation.cpp.o
[ 50%] Building CXX object src/CMakeFiles/lambda.dir/home/epics/devel/areaDetector/ADLambda/LambdaSupport/libcommon/src/ThreadUtils.cpp.o
[ 60%] Building CXX object src/CMakeFiles/lambda.dir/ImageDecoder.cpp.o
[ 70%] Building CXX object src/CMakeFiles/lambda.dir/LambdaModule.cpp.o
[ 80%] Building CXX object src/CMakeFiles/lambda.dir/LambdaSysImpl.cpp.o
[ 90%] Building CXX object src/CMakeFiles/lambda.dir/LambdaTask.cpp.o
[100%] Building CXX object src/CMakeFiles/lambda.dir/LambdaConfigReader.cpp.o
Linking CXX shared library /home/epics/devel/areaDetector/ADLambda/LambdaSupport/bin/liblambda.so
[100%] Built target lambda
corvette:ADLambda/LambdaSupport/liblambda-build>make install
[100%] Built target lambda
Install the project...
-- Install configuration: "RELEASE"
CMake Error at include/cmake_install.cmake:38 (FILE):
  file cannot create directory: /usr/local/include/lambda.  Maybe need
  administrative privileges.
Call Stack (most recent call first):
  cmake_install.cmake:37 (INCLUDE)

It is trying to install into /usr/local/include/lambda.  
Why, since i specified  
--DMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
in my cmake command above?

ulrikpedersen commented:
> Why, since i specified
> --DMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
> in my cmake command above?

Is that a typo? It should be `--DCMAKE_INSTALL_PREFIX=`

I would recommend installing the vendor software distribution separately, following the vendors instructions. Then just create appropriate macros in the build system of this module to allow users to point to the vendor library as an external dependency. See for example how we do that for the Boost, HDF5, and XML2 libraries as external dependencies... This in my opinion has a number of benefits:

 * This repo will become rather small and very much like all other drivers (i.e. easier for long term maintenance).
 * You will not have to discuss any custom build configurations with the vendor (they never like that ;-))
 * You will not have to maintain and update the vendor source code in this repository forever and ever.
 * Other users of this driver are likely to have installed the very latest version of the vendor libs+application first in order to verify operation upon delivery. I know that's the first thing that happens here when we get a new detector and it's not nice to have a slightly out-of-date version of the same library shipped with the areaDetector driver.

Just my £0.02 worth...

MarkRivers commented:
> Is that a typo? It should be --DCMAKE_INSTALL_PREFIX=

I tried changing as you suggested but it still fails the same way:

corvette:ADLambda/LambdaSupport/liblambda-build>make install
[100%] Built target lambda
Install the project...
-- Install configuration: "RELEASE"
CMake Error at include/cmake_install.cmake:38 (FILE):
  file cannot create directory: /usr/local/include/lambda.  Maybe need
  administrative privileges.
Call Stack (most recent call first):
  cmake_install.cmake:37 (INCLUDE)

I then tried 
cmake -DCMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
i.e. only 1 "-" in front of the "D" and then it worked:

corvette:ADLambda/LambdaSupport/liblambda-build>make install
[100%] Built target lambda
Install the project...
-- Install configuration: "RELEASE"
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaGlobals.h
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaInterface.h
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaSysImpl.h
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so.0.3
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so.1
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/DetInterface.h
-- Installing: /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/Globals.h

ulrikpedersen commented:
> I then tried 
> cmake -DCMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
> i.e. only 1 "-" in front of the "D" and then it worked:

Ah, yes of course. I didn't spot that one :-)

JPHammonds commented:
Sorry, yes it appears I had two typos together.
--DMAKE_INSTALL_PREFIX

-DCMAKE_INSTALL_PREFIX

I made an error translating from the INSTALL file.   I have corrected 
this and tested it out.  I have made an additional correction to the 
Makefile in the outer loop as it seems that the library version has 
changed It looks like this did not make a difference since the file 
linking  (found in LambdaSupport/liblambda-linux-x86_64/lib
lambda% ls -l
total 428
lrwxrwxrwx 1 hammonds xsdsdm     14 Apr 14 09:14 liblambda.so -> 
liblambda.so.1*
-rwxr-xr-x 1 hammonds xsdsdm 437318 Apr 14 09:14 liblambda.so.0.3*
lrwxrwxrwx 1 hammonds xsdsdm     16 Apr 14 09:14 liblambda.so.1 -> 
liblambda.so.0.3*

is lost as it is "installed" in ADLambda by the Makefile in LambdaSupport

-r--r--r-- 1 hammonds xsdsdm 843990 Apr 14 09:15 libADLambda.a
-r-xr-xr-x 1 hammonds xsdsdm 531869 Apr 14 09:15 libADLambda.so*
-r--r--r-- 1 hammonds xsdsdm 437318 Apr 14 09:15 liblambda.so
-r--r--r-- 1 hammonds xsdsdm 437318 Apr 14 09:15 liblambda.so.0.3
-r--r--r-- 1 hammonds xsdsdm 437318 Apr 14 09:15 liblambda.so.1
(After the change to the Makefile (before liblambda.so.0.3 was 
liblambda.so.0.1).

As for the Boost version problem, liblambda/INSTALL states version >= 
1.41 is required.  Which version gave you a problem?

I have not encountered the gcc problems Mark Encountered.  Since we 
needed the newer boost version we were using RHEL 7 machine.  Here I am 
using Boost 1.53 and gcc 4.8.5.  Will need to research to find where 
stdalign.h disappeared.

John





On 4/14/2016 7:25 AM, Mark Rivers wrote:
>
>     Is that a typo? It should be --DCMAKE_INSTALL_PREFIX=
>
> I tried changing as you suggested but it still fails the same way:
>
> corvette:ADLambda/LambdaSupport/liblambda-build>make install
> [100%] Built target lambda
> Install the project...
> -- Install configuration: "RELEASE"
> CMake Error at include/cmake_install.cmake:38 (FILE):
> file cannot create directory: /usr/local/include/lambda. Maybe need
> administrative privileges.
> Call Stack (most recent call first):
> cmake_install.cmake:37 (INCLUDE)
>
> I then tried
> cmake -DCMAKE_INSTALL_PREFIX=../liblambda-linux-x86_64 ../liblambda
> i.e. only 1 "-" in front of the "D" and then it worked:
>
> corvette:ADLambda/LambdaSupport/liblambda-build>make install
> [100%] Built target lambda
> Install the project...
> -- Install configuration: "RELEASE"
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaGlobals.h
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaInterface.h
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/LambdaSysImpl.h
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so.0.3
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so.1
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/lib/liblambda.so
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/DetInterface.h
> -- Installing: 
> /home/epics/devel/areaDetector/ADLambda/LambdaSupport/liblambda-linux-x86_64/include/lambda/Globals.h
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-209912864>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

MarkRivers commented:
>
As for the Boost version problem, liblambda/INSTALL states version >= 
1.41 is required. Which version gave you a problem?

1.46.0  This appears to be a known problem with that version.

>
I have not encountered the gcc problems Mark Encountered. Since we 
needed the newer boost version we were using RHEL 7 machine. Here I am 
using Boost 1.53 and gcc 4.8.5. Will need to research to find where 
stdalign.h disappeared.

Gcc 4.6.3 (my version) and 4.4.7 (the recommended version from INSTALL)  both were released in March 2012.  It appears that 4.4.7 comes with stdalign.h but 4.6.3 does not. But in any event it appears that stdalign.h is not actually needed, since it compiles without errors or warnings without it.

I just downloaded your fixes from github.  I successfully followed the instructions in INSTALL-EPICS.  I was also able to run make in LambdaSupport.  However, when running make at the top level I get the following error when building the IOC:

/usr/bin/g++ -c  -D_GNU_SOURCE -D_DEFAULT_SOURCE            -D_X86_64_  -DUNIX  -Dlinux     -O3 -g   -Wall     -D_FILE_OFFSET_BITS=64   -m64      -MMD -I. -I../O.Common -I. -I.. -I../../../include/os/Linux -I../../../include  -I/corvette/home/epics/devel/areaDetector-2-5/ADLambda/include     -I/corvette/usr/local/epics/base-3.14.12.5/include/os/Linux -I/corvette/usr/local/epics/base-3.14.12.5/include  -I/corvette/home/epics/devel/asyn-4-28/include  -I/corvette/home/epics/devel/areaDetector-2-5/ADBinaries/include  -I/corvette/home/epics/devel/areaDetector-2-5/ADCore/include -I/corvette/home/epics/devel/autosave-5-7-1/include/os/Linux -I/corvette/home/epics/devel/autosave-5-7-1/include  -I/corvette/home/epics/devel/busy-1-6-1/include  -I/corvette/home/epics/devel/calc-3-6-1/include  -I/corvette/home/epics/devel/seq-2-2-3/include  -I/corvette/home/epics/devel/sscan-2-10-2/include -I/corvette/home/epics/devel/devIocStats-3-1-14/include/os/Linux -I/corvette/home/epics/devel/devIocStats-3-1-14/include        ../LambdaAppMain.cpp 
/usr/bin/g++ -o LambdaApp -Wl,-Bstatic -L/home/epics/devel/areaDetector-2-5/ADCore/lib/linux-x86_64 -L/home/epics/devel/areaDetector-2-5/ADLambda/lib/linux-x86_64 -L/home/epics/devel/asyn-4-28/lib/linux-x86_64 -L/home/epics/devel/autosave-5-7-1/lib/linux-x86_64 -L/home/epics/devel/busy-1-6-1/lib/linux-x86_64 -L/home/epics/devel/calc-3-6-1/lib/linux-x86_64 -L/home/epics/devel/devIocStats-3-1-14/lib/linux-x86_64 -L/home/epics/devel/seq-2-2-3/lib/linux-x86_64 -L/home/epics/devel/sscan-2-10-2/lib/linux-x86_64 -L/usr/local/epics/base-3.14.12.5/lib/linux-x86_64 -L/usr/local64/lib -Wl,-rpath,/home/epics/devel/areaDetector-2-5/ADCore/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/areaDetector-2-5/ADLambda/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/asyn-4-28/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/autosave-5-7-1/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/busy-1-6-1/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/calc-3-6-1/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/devIocStats-3-1-14/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/seq-2-2-3/lib/linux-x86_64 -Wl,-rpath,/home/epics/devel/sscan-2-10-2/lib/linux-x86_64 -Wl,-rpath,/usr/local/epics/base-3.14.12.5/lib/linux-x86_64 -Wl,-rpath,/usr/local64/lib        -m64               LambdaApp_registerRecordDeviceDriver.o LambdaAppMain.o    -lADLambda -llambda -lNDPlugin -lADBase -lnetCDF -lNeXus -lcalc -lbusy -lsscan -lautosave -ldevIocStats -lasyn -lseq -lpv -lrecIoc -lsoftDevIoc -lmiscIoc -lrsrvIoc -ldbtoolsIoc -lasIoc -ldbIoc -lregistryIoc -ldbStaticIoc -lca -lCom -lhdf5 -lsz -Wl,-Bdynamic -ltiff -ljpeg -lxml2 -lz -lpthread   -lreadline -lm -lrt -ldl -lgcc
/usr/bin/ld: cannot find -llambda
collect2: ld returned 1 exit status

So it cannot find liblambda.  This is where it should find it:
corvette:~/devel/areaDetector/ADLambda>ls lib/linux-x86_64/
libADLambda.a  libADLambda.so  liblambda.so  liblambda.so.0.3  liblambda.so.1

Is the problem because I am linking statically and you have not built a static version of liblambda?  I believe this is the case because if I change this line in the IOC application Makefile
$(PROD_NAME)_LIBS += lambda
to this
$(PROD_NAME)_SYS_LIBS += lambda

it works.

JPHammonds commented:
This is probably the issue.  I usually only do static builds on 
Windows.  Since this is Linux, I am doing dynamic linking.

On 4/14/2016 9:57 AM, Mark Rivers wrote:
>
> Is the problem because I am linking statically and you have not built 
> a static version of liblambda? I believe this is the case because if I 
> change this line in the IOC application Makefile
> $(PROD_NAME)_LIBS += lambda
> to this
> $(PROD_NAME)_SYS_LIBS += lambda
>
> it works.
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

MarkRivers commented:
I think it would be safe to make the change I made.  That way it should link OK for either static or dynamic builds.  That is the way it is done for the IOC application in ADAndor, ADAndor3, and ADPointGrey.  Each of those only provides the Linux library as a .so file.  Can you test to see if you make my change that it also works for you with dynamic  build?
I just tested it by adding this line to the top of Lamda IOC application Makefile
STATIC_BUILD=NO
The resulting application was small, only 75 KB.
corvette:LambdaIOC/LambdaApp/src>ls -ltr ../../bin/linux-x86_64/
total 76
-r-xr-xr-x 1 epics epics 75091 Apr 14 10:17 LambdaApp

I then made that line be
STATIC_BUILD=YES
and the resulting application is large, 15.8 MB.
corvette:LambdaIOC/LambdaApp/src>ls -ltr ../../bin/linux-x86_64/
total 15460
-r-xr-xr-x 1 epics epics 15829128 Apr 14 10:18 LambdaApp

So it appears that using _SYS_LIBS is safe, it allows either static or dynamic builds.  Static builds are nice because I can run them on another Linux machine that does not have all of the EPICS shared libraries available.

JPHammonds commented:
OK, I will make that change.

On 4/14/2016 10:21 AM, Mark Rivers wrote:
>
> So it appears that using _SYS_LIBS is safe, it allows either static or 
> dynamic builds. Static builds are nice because I can run them on 
> another Linux machine that does not have all of the EPICS shared 
> libraries available.
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

JPHammonds commented:
OK I have made this change and pushed to the repository.

On 4/14/2016 10:21 AM, Mark Rivers wrote:
>
> I think it would be safe to make the change I made. That way it should 
> link OK for either static or dynamic builds. That is the way it is 
> done for the IOC application in ADAndor, ADAndor3, and ADPointGrey. 
> Each of those only provides the Linux library as a .so file. Can you 
> test to see if you make my change that it also works for you with 
> dynamic build?
> I just tested it by adding this line to the top of Lamda IOC 
> application Makefile
> STATIC_BUILD=NO
> The resulting application was small, only 75 KB.
> corvette:LambdaIOC/LambdaApp/src>ls -ltr ../../bin/linux-x86_64/
> total 76
> -r-xr-xr-x 1 epics epics 75091 Apr 14 10:17 LambdaApp
>
> I then made that line be
> STATIC_BUILD=YES
> and the resulting application is large, 15.8 MB.
> corvette:LambdaIOC/LambdaApp/src>ls -ltr ../../bin/linux-x86_64/
> total 15460
> -r-xr-xr-x 1 epics epics 15829128 Apr 14 10:18 LambdaApp
>
> So it appears that using _SYS_LIBS is safe, it allows either static or 
> dynamic builds. Static builds are nice because I can run them on 
> another Linux machine that does not have all of the EPICS shared 
> libraries available.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-210000115>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

MarkRivers commented:
I merged into areaDetector.  I added a comment to the IOC application Makefile about why we used _SYS_LIBS.

JPHammonds commented:
Thanks to Mark and Ulrik.

John

On 4/14/2016 10:49 AM, Mark Rivers wrote:
>
> I merged into areaDetector. I added a comment to the IOC application 
> Makefile about why we used _SYS_LIBS.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/areaDetector/ADLambda/pull/2#issuecomment-210012036>
>

-- 
--------------------------------------------
John Hammonds
Scientific Software Engineering & Data Management
Advanced Photon Source
Argonne National Laboratory
JPHammonds@anl.gov

ulrikpedersen commented:
Hi John,

You're welcome :-)

I've noticed that some of your commits are not linked to your github account. This typically happens when you make commits from a second host where your git user is not configured - or configured with an email address that you have not linked to your github account. If you do add that email address to your github account all of the commits will be updated to reference your user. Also check the local git configuration on your computers with `git config --list`

![unknown-github-user](https://cloud.githubusercontent.com/assets/6115544/14554320/1c49e374-02e4-11e6-8cee-dc48bfa804b8.png)


Cheers,
Ulrik
